File: 6197.py
import sys, heapq
input_ = sys.stdin.readline
def minput(): return map(int, input_().split())


def dijkstra(g, st):
    distances = [float('inf')] * len(g)
    distances[st] = 0
    second_dist = [float('inf')] * len(g)
    heap = []
    heapq.heappush(heap, (0, st))

    while heap:
        dist, cur = heapq.heappop(heap)
        if second_dist[cur] < dist:
            continue
        for nextnum, nextdist in g[cur]:
            t = dist + nextdist
            if distances[nextnum] > t:
                second_dist[nextnum] = distances[nextnum]
                distances[nextnum] = t
                heapq.heappush(heap, (t, nextnum))
            elif second_dist[nextnum] > t:
                second_dist[nextnum] = t
                heapq.heappush(heap, (t, nextnum))

    return second_dist


N, R = minput()
g = [[] for _ in range(N)]

for _ in range(R):
   A, B, D = minput()
   A -= 1; B -= 1
   g[A].append((B, D))
   g[B].append((A, D))

print(dijkstra(g, 0)[-1])

File: 6197.cpp
#include <iostream>
#include <vector>
#include <queue>

std::vector<std::pair<int, int> > dijkstra(const std::vector<std::vector<std::pair<int, int> > >& g, int st) {
    int INF = 2147483647;
    int N = g.size();
    std::vector<std::pair<int, int> > ret(N, std::make_pair(INF, INF));
    ret[st] = std::make_pair(0, INF);
    std::priority_queue<std::pair<int, int> > pq;
    pq.push(std::make_pair(0, st));
    while (!pq.empty()) {
        int dist = -pq.top().first;
        int cur = pq.top().second;
        pq.pop();
        if (dist > ret[cur].first) continue;
        for (const auto &edge: g[cur]) {
            int nextnum = edge.first;
            int nextdist = edge.second;
            if (ret[nextnum].first <= dist + nextdist) continue;
            ret[nextnum].first = dist + nextdist;
            pq.push(std::make_pair(-ret[nextnum].first, nextnum));
        }
    }
    return ret;
}

int main() {
    std::cin.tie(nullptr);
    std::ios_base::sync_with_stdio(false);

    int N, R;
    std::cin >> N >> R;
    std::vector<std::vector<std::pair<int, int> > > g(N);

    for (int i = 0; i < R; i++) {
        int A, B, D;
        std::cin >> A >> B >> D;
        A--;B--;
        g[A].push_back(std::make_pair(B, D));
        g[B].push_back(std::make_pair(A, D));
    }

    //
    return 0;
}
